### PR TITLE
Integrate BillKit

### PR DESCRIPTION
**Pull request by eugenijusr**
- [x] Add BillKit client library
- [x] Generate invoice after ordering a survey
- [x] Generate a refund invoice after refunding a survey
- [x] Add the ability to batch generate all the pending invoices

### CODE REVIEW

File: lib/tasks/invoices.rake [hunk 1]
```diff
@@ -0,0 +1,17 @@
+namespace :invoices do
+  desc "Create all pending invoices"
+  task :create => :environment do |task, args|
+    orders = Order.where(invoice_id: nil)
```
Comment (deviantbits) [inline lib/tasks/invoices.rake]: Maybe here, we want to order by `id ASC` to ensure the right ordering.

File: lib/tasks/invoices.rake [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    orders.each do |order|
+      CreateInvoice.perform_async(order.id)
+    end
+  end
+
+  desc "Create all pending refund invoices"
+  task :refund => :environment do |task, args|
+    orders = Order.where.not(refund: nil).where(refund_invoice_id: nil)
```
Comment (deviantbits) [inline lib/tasks/invoices.rake]: Maybe here, we want to order by `id ASC` to ensure the right ordering.

File: app/commands/create_invoice.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
+class CreateInvoice < Surveys::Command
+  def perform(order_id)
+    Honeybadger.context({
+      :order_id => order_id
+    })
+
+    order = Order.where.not(charge_token: nil).find(order_id)
+    unless order.invoice_id
+      Swagger.configure do |config|
```
Comment (deviantbits) [inline app/commands/create_invoice.rb]: `Swagger`…? :smile: Whadisdad?

File: app/commands/create_invoice.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        config.api_key = ENV['TAXAMO_SECRET']
+      end
+
+      response = Taxamo.get_transaction(order.charge_token)
+      if response.transaction
+        transaction_line = response.transaction.transaction_lines.first
+
+        client = BillKit::Client.new(ENV['BILLKIT_KEY'], ENV['BILLKIT_SECRET'])
```
Comment (deviantbits) [inline app/commands/create_invoice.rb]: This is *beautiful*.

File: app/commands/create_invoice.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        invoice = client.create_invoice(
+          order.email,
+          [{
+            title: "Participants for survey \"#{order.survey.title}\"",
```
Comment (deviantbits) [inline app/commands/create_invoice.rb]: I'd use `Participant` here (singular).

File: app/commands/create_invoice.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            unit_price: (order.price.to_f / order.requested_participant_count.to_f).round,
+            quantity: order.requested_participant_count,
+            tax_rate: transaction_line.tax_rate,
+            supplied_at: order.charged_at
+          }],
+          {
+            vat_number: response.transaction.buyer_tax_number,
```
Comment (deviantbits) [inline app/commands/create_invoice.rb]: Wow. So clean, so awesome.

File: app/commands/create_refund_invoice.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      response = Taxamo.get_transaction(order.charge_token)
+      if response.transaction
+        transaction_line = response.transaction.transaction_lines.first
+
+        client = BillKit::Client.new(ENV['BILLKIT_KEY'], ENV['BILLKIT_SECRET'])
+        invoice = client.refund_invoice(
+          order.invoice_id,
+          {
+            items: [{
+              title: "Participants for survey \"#{order.survey.title}\"",
```
Comment (deviantbits) [inline app/commands/create_refund_invoice.rb]: Again, I'd rather say `Participant` here.

File: app/commands/refund.rb [hunk 8]
```diff
@@ -41,6 +41,9 @@ def perform(order_id)
       order.refund = ((taxamo_refund.total_amount - (taxamo_refund.tax_amount || 0)) * 100).round
       order.save!
 
+      # Create refund invoice.
```
Comment (deviantbits) [inline app/commands/refund.rb]: I'd remove the comments. They're redundant and useless. See http://apdevblog.com/comments-in-code/.

Comment (deviantbits) [inline app/commands/refund.rb]:
```
# Create refund invoice.
CreateRefundInvoice.perform_async
```
Come on… :wink:

File: app/views/orders/_taxamo.html.erb [hunk 9]
```diff
@@ -10,13 +10,14 @@
     $("#new_order").validate();
     if($('#new_order').valid() === true) {
       var participants = $('#order_requested_participant_count').val();
-      var amount = $('#order_duration').val() * 10 * participants;
+      var price = $('#order_duration').val() * 10;
```
Comment (deviantbits) [inline app/views/orders/_taxamo.html.erb]: Much nicer.

**Comment by deviantbits** [comment]
I've reviewed this. Do you need any `de` translations?

File: app/commands/create_invoice.rb (hunk 3 above)
Comment (eugenijusr) [inline app/commands/create_invoice.rb]: This is how Taxamo is configured :)

**Comment by eugenijusr** [comment]
Thanks for the review!

Regarding localization. Since invoices & emails are not localized in BillKit at this point I think there's no need to localize the invoice line `Participant for survey {x}` as well. We should make BillKit localizable in the future however. By the way is it OK that all the invoices are in English right now?

**Comment by deviantbits** [comment]
Yup. I deliberately did this all in English. This should be fine. If I hear from our tax agent otherwise, I'll implement the necessary localizations later. (In that case, we'd have to regenerate.) – But this might also happen in other cases, so I'd like to try this first.

**Pull request merged**